File: Plugins/UnrealMCP/cli/cmd/debug.go
package cmd

func init() {
	ensureGroup("debug", "Debug")
	registerCommands(debugCommands)
}

var debugCommands = []CommandSpec{
	{
		Name:    "set_mcp_debug",
		Group:   "debug",
		Short:   "Enable or disable MCP debug mode (token tracking)",
		Long:    "Toggles MCP debug mode which enables token usage tracking on every MCP command. When enabled, the bridge measures the approximate token size of each request and response, accumulates totals per command, and tracks call counts. Use get_mcp_token_stats to read the accumulated statistics. Useful for identifying which commands produce oversized responses that waste LLM context budget.",
		Example: `ue-cli set_mcp_debug --enabled
ue-cli set_mcp_debug --enabled=false`,
		Params: []ParamSpec{
			{Name: "enabled", Type: "bool", Default: true, Help: "Enable debug mode"},
		},
	},
	{
		Name:    "get_mcp_token_stats",
		Group:   "debug",
		Short:   "Get MCP token usage statistics",
		Long:    "Returns accumulated token usage statistics from the MCP bridge. Shows per-command call counts, total input/output tokens, and averages. Only collects data when debug mode is enabled via set_mcp_debug. Use this to audit which MCP commands consume the most tokens and optimize your workflow accordingly.",
		Example: "ue-cli get_mcp_token_stats",
	},
}

File: Plugins/UnrealMCP/unrealmcp/tools/debug.py
"""Debug tools — token estimation, response size analysis."""

from unrealmcp._bridge import mcp
from unrealmcp._tcp_bridge import _call


@mcp.tool()
def set_mcp_debug(enabled: bool = True) -> str:
    """Enable or disable token debug estimation on MCP responses.

    When enabled, every command response shows a token debug header:
        [TOKEN DEBUG] command_name — ~1,289 tokens (5,156 chars)

    This appears at the top of every response so you can immediately
    see which commands consume the most tokens.

    Disabling clears accumulated stats.

    Args:
        enabled: True to enable, False to disable and clear stats.
    """
    return _call("set_mcp_debug", {"enabled": enabled})


@mcp.tool()
def get_mcp_token_stats() -> str:
    """Get accumulated per-command token usage statistics.

    Returns commands sorted by max_estimated_tokens (worst offenders first).
    Each entry shows: call_count, avg/max/total estimated tokens, response chars.

    Enable debug mode first with set_mcp_debug(enabled=True).
    """
    return _call("get_mcp_token_stats")
